pepper-apis/docs/specification/src/components/schemas.yml

```diff
@@ -1720,6 +1795,56 @@ Question.Picklist:
         type: array
         items:
           $ref: '#/Answer.Picklist'
+Question.Matrix:
+  allOf:
+    - $ref: '#/Question'
+    - type: object
+      required:
+        - selectMode
+        - matrixOptions
+        - matrixQuestions
+        - answers
+      properties:
+        questionType:
+          type: string
+          default: MATRIX
+          enum:
+            - MATRIX
+        selectMode:
+          type: string
+          description: |
+            * `MULTIPLE` - Allows user to select as many options as needed (checkbox)
+            * `SINGLE` - Only one selection is allowed inside group (radio button)
+          enum:
+            - MULTIPLE
+            - SINGLE
+        groups:
+
+          type: array
+          description: A list of groups used in the `matrixOptions` values. If there are no options which explicity request grouping, this will be an empty array.
+          minItems: 0
+          items:
+            type: object
+            required:
+              - identifier
+              - name
+            properties:
+              identifier:
+                type: string
```

Comment:
Is your intent here to prevent ungrouped (`MatrixOption` objects with a null `groupId`) from having a prompt?

Reply:
Kind of, if option has no group, there will be no additional header over it

Reply:
Is there a reason we're special casing the ungrouped options like that? If not, I'd prefer that we treat everything the same- how difficult a change would it be to require that every option belong to a group _or_ allow a header to be declared for the `null` group?

Reply:
Well, most of the times in CMI there's no group usage at all (51 vs 1). It's just a concept that allow us to separate some columns and it does the thing. It's a whole lot of options to choose: without groups (null group only), null group + some groups, groups only. I believe it's quite flexible.

Reply:
Also, in the given example, there's no need for additional header over the `null` group. If we were in need of additional header, I'd just declare a group for it and that's it.
<img width="522" alt="image" src="https://user-images.githubusercontent.com/71830706/144391726-2bca8891-3a40-4150-96b1-ced3bd6f8865.png">

Reply:
> It's a whole lot of options to choose
Yes, that is my point. It's flexibility at the cost of client complexity, and doesn't seem be flexibility that is needed at this point. If an implementer wants to use groups, or needs the features provided by them, everything should be in a group. Otherwise, they can leave the `groupId` set to `null` and not opt-in to the extra behaviors.

Reply:
I'm not sure I get what the client complexity is. `null` group is simply treated as a group without any additional header. @aleksey-kuznetsov you have written some code for that on the SDK side. What are your thoughts on the above?

Reply:
From the UI perspective every possible case is covered and I don't see any complexities since ungrouped options (whose `groupId` is set to `null`) are actually a part of `null` group. So technically each and every option is a part of a group and there is no case like `ungrouped options` + `grouped options`.

Reply:
And correct, for `null` group there is no header text and it cannot be configured. If one is needed -- just group these options in a named group and set a header for it.

Reply:
Using a `null` group id as short-hand for "This group doesn't have a header" is not a logical grouping of options- it is a UI treatment leaking into the API structure.

For me to feel comfortable approving this PR, I would like the following changes:

* `MatrixGroup.name` should be `nullable`
* If any groups are used, _all_ options should belong to a declared group
* If any options have a `null` `groupId`, all options should have a `null` group id.

@MocanaAtBroad or @ssettipalli , if you feel these changes are unnecessary, please dismiss my review and approve this PR at your discretion.

Reply:
> Using a null group id as short-hand for "This group doesn't have a header"
I see it from different angle. As I said, most of the times there shouldn't be any groups at all (51 question VS 1 question with groups for CMI). However, sometimes we need additional headers AND separate selection within some set of options and this is achieved by these groups. Most of the times there shouldn't be any groups at all.

Reply:
In other words, using group is short-hand for "This group should have a separate selection even if select mode is SINGLE, and it should have additional header so that user see that this set of options is separate from the rest"

Reply:
- `MatrixGroup.name` should be `nullable`
With this change there will be only one usage of a group -- if we need an option to be `exclusive` among some specific options and not all of them. Since groups are not highlighted in any way (unless you style it on per study basis) it can be confusing without a name of a group to identify what options belong where.

As for the points #2 and #3 a `groupId` of `null` is a shorthand for "This option is in unnamed group" so it not only doesn't have a header but all other specifics like `exclusive` flag will work within this unnamed group.